lib/IntermediateRepresentation/basicblock.h

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef FORTRAN_INTERMEDIATEREPRESENTATION_BASICBLOCK_H_
+#define FORTRAN_INTERMEDIATEREPRESENTATION_BASICBLOCK_H_
+
+#include "mixin.h"
+#include "region.h"
+#include <iostream>
+
+namespace Fortran::IntermediateRepresentation {
+
+struct Region;
+struct Statement;
+
+struct BasicBlock final : public llvm::ilist_node<BasicBlock>,
+                          public ChildMixin<BasicBlock, Region> {
+  using StatementListType = llvm::iplist<Statement>;
+  using iterator = StatementListType::iterator;
+  using const_iterator = StatementListType::const_iterator;
+  using reverse_iterator = StatementListType::reverse_iterator;
+  using const_reverse_iterator = StatementListType::const_reverse_iterator;
+
+  BasicBlock(const BasicBlock &) = delete;
+  BasicBlock &operator=(const BasicBlock &) = delete;
+  ~BasicBlock();
+  StatementListType &getSublist(Statement *) { return Statements(); }
+  void insertBefore(Statement *stmt, Statement *before = nullptr);
```

Comment:
Where you have pointer arguments to functions that cannot be null, consider making them references instead.

Reply:
Will do.  I have used `Foo *` as scaffolding in some contexts in this prototype-ish code since I hadn't zeroed in on exactly what bits a `Foo` reifies to yet.